### ISSUE TITLE
Kind annotations should be stripped from IDE responses

### ISSUE THREAD

**Issue by thomashoneyman** [purescript/purescript#4037]
## Description

As of PureScript 0.14 I've noticed that autocomplete suggestions via the language server have become noisier because they now include kind information. I don't think that this information is particularly helpful (in most cases, at least), and it makes some functions nearly impossible to understand from the suggestion window alone.

For example, the [`mapReaderT`](https://github.com/purescript/purescript-transformers/blob/846581729493343f5257d62c142ab5d1746b18dc/src/Control/Monad/Reader/Trans.purs#L42-L43) function from `transformers`:

```purs
mapReaderT :: forall r m1 m2 a b. (m1 a -> m2 b) -> ReaderT r m1 a -> ReaderT r m2 b
```

has this output as its autocomplete suggestion:

<img width="425" alt="Screen Shot 2021-03-24 at 8 48 08 PM" src="https://user-images.githubusercontent.com/10245104/112416494-06886900-8ce3-11eb-86ed-7873e0371b5d.png">

Not only do the kind annotations make things busier, we've even got two type variables not otherwise included in the type definition. Here's another example -- the [`runUI`](https://github.com/purescript-halogen/purescript-halogen/blob/e6ac20be9531356d9c01fd8cab9e3ddf87c27c9a/src/Halogen/VDom/Driver.purs#L129-L134) function from `halogen`:

```purs
runUI
  :: forall query input output
   . Component HTML query input output Aff
  -> input
  -> DOM.HTMLElement
  -> Aff (HalogenIO query output Aff)
```

Admittedly, `HalogenIO` is itself a record that is desugared in this output. Even so, you can see how the kind information on top of the desugaring makes it nearly impossible to understand what's going on. I even have to scroll!

<img width="424" alt="Screen Shot 2021-03-24 at 8 41 00 PM" src="https://user-images.githubusercontent.com/10245104/112416624-45b6ba00-8ce3-11eb-801b-84471415c1a2.png">

## To Reproduce

Using the autosuggest feature in an editor using the language server should find plenty of examples, but the two above can be good places to start.

## Expected behavior

The autosuggested type should not include kind information, or at least should only include kind information that was explicitly provided by the author, or at least should only include the standalone kind signature if there was one.

## PureScript version

0.14.0

**Comment by hdgarrood** [comment]
I think removing kind information entirely would be going too far; it’s exactly what you’d want to see in this context if you’re doing type level programming, and there’s an important difference between, for example, `Proxy :: Type -> Type` and `Proxy :: forall k. k -> Type`. However it does look like there is some compiler-internal information which ought not to be there in the mapReaderT output.

I’m less sure about what the ide server should do for runUI. What do you get when you ask for its type in 0.13.x?

**Comment by thomashoneyman** [comment]
In PureScript 0.13.x, `mapReaderT`:

<img width="394" alt="Screen Shot 2021-03-25 at 10 23 54 AM" src="https://user-images.githubusercontent.com/10245104/112516211-48052c80-8d54-11eb-8fdb-6493ef557693.png">

and `runUI`:

<img width="395" alt="Screen Shot 2021-03-25 at 10 23 42 AM" src="https://user-images.githubusercontent.com/10245104/112516238-4d627700-8d54-11eb-8762-14f10a8bf68f.png">

`runUI` still isn't fantastic here, but it's not quite as gnarly as it is with the kind information too.

> I think removing kind information entirely would be going too far; it’s exactly what you’d want to see in this context if you’re doing type level programming, and there’s an important difference between, for example, `Proxy :: Type -> Type` and `Proxy :: forall k. k -> Type`.

I wonder if there are any heuristics we could use to determine whether to include the kind information -- for example, include it if there are polymorphic kinds, or include it if there are explicit user-provided annotations? Does anything come to mind that would help provide the information in the cases where it is most likely to be useful, and not provide it otherwise?

**Comment by natefaubion** [comment]
I think we should probably use the same routine we use when printing (non-verbose) errors.
https://github.com/purescript/purescript/blob/ee0b3d391151bcd5f56de4563208dcf657cccc8c/src/Language/PureScript/Errors.hs#L1591

**Comment by natefaubion** [comment]
I think ideally we would just retain the original source annotation and show that, but that is not in externs.

**Comment by kritzcreek** [comment]
> I think ideally we would just retain the original source annotation and show that, but that is not in externs.

FWIW we do that when the provided source globs contain the file in question. The IDE server returns both, and I think the VSCode plugin chooses to show them both in the popup, if they differ:

![image](https://user-images.githubusercontent.com/6189397/112522407-e1a4fc00-8d9d-11eb-9f5e-8336db1c5c26.png)

The most likely cause for not getting the user provided signature is not having checked the "Add spago sources" box in the VSCode plugin's settings.

**Comment by natefaubion** [comment]
Ah, great point! I'm less inclined to worry about this then.

> However it does look like there is some compiler-internal information which ought not to be there in the mapReaderT output.

I think the only thing would be kind applications, which are easy to remove, but while it's "internal", it's the only way to observe how such a thing is being instantiated. I'm not sure it's any more internal that skolems or unknowns, which invariably leak out in sometimes odd ways.

**Comment by thomashoneyman** [comment]
@kritzcreek Good point; I neglected to mention that for this project I was using Bower and Pulp. Using Spago on a different project I do see a much better output for `runUI`:

<img width="397" alt="Screen Shot 2021-03-26 at 10 22 26 AM" src="https://user-images.githubusercontent.com/10245104/112669779-52d7c400-8e1d-11eb-941b-bf63f2276ddf.png">

**Comment by f-f** [comment]
I see both signatures even if I have the "use Spago sources" checkbox on, but this might just be an issue with my VSCode setup.

**Comment by milesfrain** [comment]
Should kind annotations be included in REPL `:type` responses? This can make things more confusing (or at least busier) for beginners as noted in https://github.com/purescript-contrib/purescript-book/issues/326.
